Fix clang-format script behaviour without -i + code cleanup

oxan opened the pull request:
# What does this implement/fix? 

Primarily this changes `script/clang-format` to actually output the formatting problems when invoked without the `-i` flag. Since the upgrade to clang-11 it needs either the `-i` or `-n -Werror` flags, otherwise it does nothing at all. 

Also make the exit code non-zero if any files need to be changed, and some code cleanup.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**  N/A

## Test Environment

- [ ] ESP32
- [ ] ESP8266

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

oxan committed:
Drop Python 2 support

oxan committed:
Drop unused code

oxan committed:
Make clang-format output files to be changed

Also ensures exit code is non-zero if any files need to be changed.

oxan committed:
Clean-up clang-format script

oxan committed:
Clean-up clang-tidy script

agners commented:
> Since the upgrade to clang-11 it needs either the `-i` or `-n -Werror` flags, otherwise it does nothing at all.

Technically `clang-format` writes the formatted file to stdout by default (both back with version 7 and 11) and exits with exit code 0. However, we don't do anything with the output in that case, so the script ends up doing nothing essentially. That was the case with clang-format-7 as well though.

With clang-format-11 now has the `-n` option (or `--dry-run`, which I actually prefer in scripts as it makes it clear what the options actually means). With your change, defaulting `script/clang-format` to use dry-run mode and print all messages to the console, invocation without -i actually does something sensible.

I agree with your change, just wanted to clarify a bit :)

agners commented on script/clang-format:
```diff
@@ -13,59 +12,47 @@ import threading
         if args.inplace:
             invocation.append('-i')
+        else:
+            invocation.extend(['-n', '-Werror'])
```
Nit
```suggestion
            invocation.extend(['--dry-run', '-Werror'])
```

oxan committed:
Update script/clang-format

Co-authored-by: Stefan Agner <stefan@agner.ch>

oxan commented:
> That was the case with clang-format-7 as well though.

Ah, you're right. I thought that something did happen while running without `-i` with clang-7, but I just checked and it seems I remembered wrong.

jesserockz commented:
A non-zero exit code will break the CI flow as it will not be able to run the following `script/ci-suggest-changes` script.

Can you make sure it "works" by adding some bad formatting to this PR? (Can remove after)

agners commented:
@jesserockz indeed this change changes the behavior of `script/clang-format` wrt exit code.

It seems clang-tidy already exits non-zero and doesn't call the `script/ci-suggest-changes` script, see:
https://github.com/esphome/esphome/pull/2030/checks?check_run_id=3087670070

I doubt that this is intentional.

IMHO returning non-zero makes sense for `script/clang-format`. I think a change like this to make CI continue makes sense:
```
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -48,10 +48,12 @@ jobs:
 
       - name: Run clang-format
         run: script/clang-format -i
+        continue-on-error: true
         if: ${{ matrix.id == 'clang-format' }}
 
       - name: Run clang-tidy
         run: script/clang-tidy --all-headers --fix --split-num 4 --split-at ${{ matrix.split }}
+        continue-on-error: true
         if: ${{ matrix.id == 'clang-tidy' }}
 
       - name: Suggest changes
```

jesserockz commented:
Yeah that would be fine then

oxan commented:
@agners The issue with that is that clang-tidy doesn't have fixers for all issues it detects, so it could fail but also leave the worktree unchanged, which would result in the job succeeding while there are issues.

jesserockz commented:
We can add another step then that can check the success status of the clang-tidy job and fail the job (after suggest changes if that passes)

oxan commented:
Ah, didn't know that was possible, sounds like a path forward (though I can't help but laugh about how complex this is becoming :laughing:)

agners commented:
There is a simpler way to do this:

```
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -56,6 +56,7 @@ jobs:
 
       - name: Suggest changes
         run: script/ci-suggest-changes
+        if: always()
 
   ci:
     # Don't use the esphome-lint docker image because it may contain outdated requirements.

```

Note the file has changed in `dev` compared to your branch, so you have to rebase/merge.

agners commented:
Btw, I tested that on my end, to make sure it does what I think it should :)
https://github.com/agners/esphome/runs/3132089944?check_suite_focus=true

oxan committed:
Always run suggest changes CI step

Co-authored-by: Stefan Agner <stefan@agner.ch>

oxan committed:
Mark clang-format step as failed if it changed files

oxan commented:
@agners Thanks, that's a really nice solution. I've rebased and added it to the PR.

@jesserockz See [this run](https://github.com/oxan/esphome/actions/runs/1056875213) as an example of a failed run (it's this branch + [one commit](https://github.com/oxan/esphome/commit/6e8c70731a3a5d231ed3c84a485c1f9b4d8144be) to introduce errors).
